File: routes/farmStaff.js
const express = require('express');
const router = express.Router();
const { User, UserFarm, Farm } = require('../models');
const { isAuthenticated } = require('../middleware/auth');
const { Op } = require('sequelize');

// 권한 체크: system_admin, super_admin, 또는 해당 농장의 farm_admin 만 직원 관리 가능
async function checkStaffPermission(req, res, next) {
    const { farmId } = req.params;

    if (!farmId) {
        return res.status(400).json({ error: 'farmId가 필요합니다.' });
    }

    if (['system_admin', 'super_admin'].includes(req.session.systemRole)) {
        return next();
    }

    try {
        const uf = await UserFarm.findOne({
            where: {
                userId: req.session.userId,
                farmId,
                role: 'farm_admin',
                isActive: true
            }
        });

        if (!uf) {
            return res.status(403).json({ error: '해당 농장의 직원을 관리할 권한이 없습니다.' });
        }

        return next();
    } catch (error) {
        console.error('직원 권한 확인 오류:', error);
        return res.status(500).json({ error: '서버 오류가 발생했습니다.' });
    }
}

// 특정 농장의 직원 목록
router.get('/:farmId/staff', isAuthenticated, checkStaffPermission, async (req, res) => {
    try {
        const { farmId } = req.params;

        const staff = await UserFarm.findAll({
            where: { farmId, isActive: true },
            include: [{
                model: User,
                as: 'user',
                attributes: ['id', 'username', 'fullName', 'email', 'phone', 'systemRole']
            }],
            order: [['createdAt', 'ASC']]
        });

        // system_admin, super_admin 은 직원 목록에서 제외
        const filtered = staff.filter(uf => !['system_admin', 'super_admin'].includes(uf.user?.systemRole));

        const result = filtered.map(uf => ({
            userFarmId: uf.id,
            userId: uf.userId,
            username: uf.user?.username,
            fullName: uf.user?.fullName,
            email: uf.user?.email,
            phone: uf.user?.phone,
            role: uf.role,
            department: uf.department,
            position: uf.position,
            employmentType: uf.employmentType,
            hireDate: uf.hireDate,
            resignDate: uf.resignDate,
            isActive: uf.isActive
        }));

        res.json(result);
    } catch (error) {
        console.error('직원 목록 조회 오류:', error);
        res.status(500).json({ error: '직원 목록을 불러오는 중 오류가 발생했습니다.' });
    }
});

// 새 계정 + 직원 등록
router.post('/:farmId/staff', isAuthenticated, checkStaffPermission, async (req, res) => {
    const { farmId } = req.params;
    const { account, staff } = req.body || {};

    try {
        if (!account || !account.username || !account.password || !account.fullName) {
            return res.status(400).json({ error: '계정 정보(아이디, 비밀번호, 이름)는 필수입니다.' });
        }

        const existing = await User.findOne({ where: { username: account.username } });
        if (existing) {
            return res.status(400).json({ error: '이미 사용 중인 아이디입니다.' });
        }

        const user = await User.create({
            username: account.username,
            password: account.password, // TODO: 비밀번호 해시 적용 필요 (auth 라우트와 동일 로직 재사용 권장)
            fullName: account.fullName,
            phone: account.phone || null,
            email: (account.email && String(account.email).trim()) || null,
            birthDate: account.birthDate || null,
            gender: account.gender || null,
            nationalityType: account.nationalityType || 'domestic',
            nationality: account.nationality || null,
            visaType: account.visaType || null,
            visaExpiry: account.visaExpiry || null,
            postalCode: account.postalCode || null,
            address: account.address || null,
            addressDetail: account.addressDetail || null,
            systemRole: 'user'
        });

        const role = staff?.role || 'staff';

        const userFarm = await UserFarm.create({
            userId: user.id,
            farmId,
            role,
            department: staff?.department || null,
            position: staff?.position || null,
            employmentType: staff?.employmentType || 'full_time',
            hireDate: staff?.hireDate || null,
            salary: staff?.salary || null,
            assignedBy: req.session.userId,
            assignedAt: new Date()
        });

        res.status(201).json({ user, userFarm });
    } catch (error) {
        console.error('직원 등록 오류:', error);
        res.status(500).json({ error: '직원 등록 중 오류가 발생했습니다.' });
    }
});

// 직원(농장 소속) 수정
router.put('/:farmId/staff/:userFarmId', isAuthenticated, checkStaffPermission, async (req, res) => {
    const { farmId, userFarmId } = req.params;
    const { user: userBody, staff: staffBody } = req.body || {};

    try {
        const uf = await UserFarm.findOne({
            where: { id: userFarmId, farmId, isActive: true },
            include: [{ model: User, as: 'user', attributes: ['id', 'fullName', 'phone', 'email'] }]
        });
        if (!uf) {
            return res.status(404).json({ error: '해당 농장의 직원 정보를 찾을 수 없습니다.' });
        }

        if (staffBody) {
            if (staffBody.role != null) uf.role = staffBody.role;
            if (staffBody.department !== undefined) uf.department = staffBody.department || null;
            if (staffBody.position !== undefined) uf.position = staffBody.position || null;
            if (staffBody.employmentType != null) uf.employmentType = staffBody.employmentType;
            if (staffBody.hireDate !== undefined) uf.hireDate = staffBody.hireDate || null;
            if (staffBody.resignDate !== undefined) uf.resignDate = staffBody.resignDate || null;
            if (staffBody.salary !== undefined) uf.salary = staffBody.salary || null;
            await uf.save();
        }

        if (userBody && uf.user) {
            const user = await User.findByPk(uf.userId);
            if (user) {
                if (userBody.fullName !== undefined) user.fullName = userBody.fullName;
                if (userBody.phone !== undefined) user.phone = userBody.phone || null;
                if (userBody.email !== undefined) user.email = (userBody.email && String(userBody.email).trim()) || null;
                await user.save();
            }
        }

        const updated = await UserFarm.findByPk(uf.id, {
            include: [{ model: User, as: 'user', attributes: ['id', 'username', 'fullName', 'email', 'phone'] }]
        });
        res.json(updated);
    } catch (error) {
        console.error('직원 수정 오류:', error);
        res.status(500).json({ error: '직원 수정 중 오류가 발생했습니다.' });
    }
});

// 직원(농장 소속) 삭제 — 퇴사 처리(soft: isActive=false, resignDate=오늘)
router.delete('/:farmId/staff/:userFarmId', isAuthenticated, checkStaffPermission, async (req, res) => {
    const { farmId, userFarmId } = req.params;

    try {
        const uf = await UserFarm.findOne({ where: { id: userFarmId, farmId } });
        if (!uf) {
            return res.status(404).json({ error: '해당 농장의 직원 정보를 찾을 수 없습니다.' });
        }
        uf.isActive = false;
        uf.resignDate = uf.resignDate || new Date().toISOString().slice(0, 10);
        await uf.save();
        res.json({ success: true, message: '퇴사 처리되었습니다.' });
    } catch (error) {
        console.error('직원 삭제(퇴사) 오류:', error);
        res.status(500).json({ error: '직원 삭제 처리 중 오류가 발생했습니다.' });
    }
});

// 기존 계정(다른 농장 직원 포함)을 이 농장 직원으로 연결
router.post('/:farmId/staff/link', isAuthenticated, checkStaffPermission, async (req, res) => {
    const { farmId } = req.params;
    const { userId, role, department, position, employmentType, hireDate, salary } = req.body || {};

    try {
        if (!userId) {
            return res.status(400).json({ error: 'userId는 필수입니다.' });
        }

        const user = await User.findByPk(userId);
        if (!user) {
            return res.status(404).json({ error: '해당 사용자를 찾을 수 없습니다.' });
        }

        const existing = await UserFarm.findOne({ where: { userId, farmId } });
        if (existing) {
            return res.status(400).json({ error: '이미 이 농장에 등록된 직원입니다.' });
        }

        const userFarm = await UserFarm.create({
            userId,
            farmId,
            role: role || 'staff',
            department: department || null,
            position: position || null,
            employmentType: employmentType || 'full_time',
            hireDate: hireDate || null,
            salary: salary || null,
            assignedBy: req.session.userId,
            assignedAt: new Date()
        });

        res.status(201).json({ user, userFarm });
    } catch (error) {
        console.error('기존 직원 연결 오류:', error);
        res.status(500).json({ error: '기존 직원을 이 농장에 추가하는 중 오류가 발생했습니다.' });
    }
});

module.exports = router;

// 직원 검색 (기존 계정/다른 농장 직원 포함)
// 경로: /api/farms/staff/search?query=...
router.get('/staff/search', isAuthenticated, async (req, res) => {
    try {
        const { query, farmId } = req.query;

        // system_admin, super_admin 만 전체 검색 허용
        if (!['system_admin', 'super_admin'].includes(req.session.systemRole)) {
            return res.status(403).json({ error: '직원 검색 권한이 없습니다.' });
        }

        const whereUser = {};
        if (query && query.trim()) {
            const q = `%${query.trim()}%`;
            whereUser[Op.or] = [
                { username: { [Op.iLike]: q } },
                { fullName: { [Op.iLike]: q } },
                { phone: { [Op.iLike]: q } },
                { email: { [Op.iLike]: q } }
            ];
        }

        const users = await User.findAll({
            where: whereUser,
            include: [{
                model: UserFarm,
                as: 'userFarms',
                include: [{
                    model: Farm,
                    as: 'farm',
                    attributes: ['id', 'farmName']
                }]
            }],
            order: [['createdAt', 'DESC']],
            limit: 30
        });

        const currentFarmId = farmId || null;

        const results = users.map(user => {
            const farms = (user.userFarms || []).map(uf => ({
                farmId: uf.farmId,
                farmName: uf.farm ? uf.farm.farmName : null,
                role: uf.role
            }));

            const isInCurrentFarm = currentFarmId
                ? farms.some(f => f.farmId === currentFarmId)
                : false;

            return {
                userId: user.id,
                username: user.username,
                fullName: user.fullName,
                phone: user.phone,
                email: user.email,
                farms,
                isInCurrentFarm
            };
        });

        res.json(results);
    } catch (error) {
        console.error('직원 검색 오류:', error);
        res.status(500).json({ error: '직원 검색 중 오류가 발생했습니다.' });
    }
});

File: backend/internal/handlers/farm_staff.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"

	"pig-farm-api/internal/middleware"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
)

// FarmStaffList GET /api/farms/:farmId/staff
func (h *Handler) FarmStaffList(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r.Context())
	if claims == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "로그인이 필요합니다."})
		return
	}
	farmIDStr := chi.URLParam(r, "farmId")
	farmID, err := uuid.Parse(farmIDStr)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "farmId가 필요합니다."})
		return
	}
	if !h.canManageFarmStaff(r.Context(), claims, farmID) {
		writeJSON(w, http.StatusForbidden, map[string]string{"error": "해당 농장의 직원을 관리할 권한이 없습니다."})
		return
	}
	rows, err := h.db.Pool.Query(r.Context(), `
		SELECT uf.id, uf."userId", u.username, u."fullName", u.email, u.phone, u."systemRole", uf.role, uf.department, uf.position,
		       uf."employmentType", uf."hireDate", uf."resignDate", uf."isActive"
		FROM user_farms uf
		JOIN users u ON u.id = uf."userId"
		WHERE uf."farmId" = $1 AND uf."isActive" = true AND u."systemRole" NOT IN ('super_admin', 'system_admin')
		ORDER BY uf."createdAt" ASC
	`, farmID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "직원 목록을 불러오는 중 오류가 발생했습니다."})
		return
	}
	defer rows.Close()
	type staffRow struct {
		UserFarmID    string  `json:"userFarmId"`
		UserID        string  `json:"userId"`
		Username      *string `json:"username"`
		FullName      *string `json:"fullName"`
		Email         *string `json:"email"`
		Phone         *string `json:"phone"`
		Role          string  `json:"role"`
		Department    *string `json:"department"`
		Position      *string `json:"position"`
		EmploymentType *string `json:"employmentType"`
		HireDate      *string `json:"hireDate"`
		ResignDate    *string `json:"resignDate"`
		IsActive      bool   `json:"isActive"`
	}
	var list []staffRow
	for rows.Next() {
		var ufID, uID uuid.UUID
		var username, fullName, email, phone interface{}
		var systemRole string
		var role string
		var dept, pos, empType, hireDate, resignDate interface{}
		var isActive bool
		if err := rows.Scan(&ufID, &uID, &username, &fullName, &email, &phone, &systemRole, &role, &dept, &pos, &empType, &hireDate, &resignDate, &isActive); err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "직원 목록을 불러오는 중 오류가 발생했습니다."})
			return
		}
		s := staffRow{UserFarmID: ufID.String(), UserID: uID.String(), Role: role, IsActive: isActive}
		if v, ok := username.(string); ok {
			s.Username = &v
		}
		if v, ok := fullName.(string); ok {
			s.FullName = &v
		}
		if v, ok := email.(string); ok {
			s.Email = &v
		}
		if v, ok := phone.(string); ok {
			s.Phone = &v
		}
		if v, ok := dept.(string); ok {
			s.Department = &v
		}
		if v, ok := pos.(string); ok {
			s.Position = &v
		}
		if v, ok := empType.(string); ok {
			s.EmploymentType = &v
		}
		if v, ok := hireDate.(string); ok {
			s.HireDate = &v
		}
		if v, ok := resignDate.(string); ok {
			s.ResignDate = &v
		}
		list = append(list, s)
	}
	writeJSON(w, http.StatusOK, list)
}

// FarmStaffCreate POST /api/farms/:farmId/staff (new user + user_farm)
func (h *Handler) FarmStaffCreate(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r.Context())
	if claims == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "로그인이 필요합니다."})
		return
	}
	farmIDStr := chi.URLParam(r, "farmId")
	farmID, err := uuid.Parse(farmIDStr)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "farmId가 필요합니다."})
		return
	}
	if !h.canManageFarmStaff(r.Context(), claims, farmID) {
		writeJSON(w, http.StatusForbidden, map[string]string{"error": "해당 농장의 직원을 관리할 권한이 없습니다."})
		return
	}
	var body struct {
		Account struct {
			Username  string  `json:"username"`
			Password  string  `json:"password"`
			FullName  string  `json:"fullName"`
			Phone     *string `json:"phone"`
			Email     *string `json:"email"`
		} `json:"account"`
		Staff struct {
			Role           *string `json:"role"`
			Department     *string `json:"department"`
			Position       *string `json:"position"`
			EmploymentType *string `json:"employmentType"`
			HireDate       *string `json:"hireDate"`
		} `json:"staff"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청입니다."})
		return
	}
	if body.Account.Username == "" || body.Account.Password == "" || body.Account.FullName == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "계정 정보(아이디, 비밀번호, 이름)는 필수입니다."})
		return
	}
	var exists int
	if err := h.db.Pool.QueryRow(r.Context(), `SELECT 1 FROM users WHERE username = $1`, body.Account.Username).Scan(&exists); err == nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "이미 사용 중인 아이디입니다."})
		return
	}
	userID := uuid.New()
	_, err = h.db.Pool.Exec(r.Context(), `
		INSERT INTO users (id, username, password, "fullName", phone, email, "systemRole", "createdAt", "updatedAt")
		VALUES ($1, $2, $3, $4, $5, $6, 'user', NOW(), NOW())
	`, userID, body.Account.Username, body.Account.Password, body.Account.FullName, body.Account.Phone, body.Account.Email)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "직원 등록 중 오류가 발생했습니다."})
		return
	}
	role := "staff"
	if body.Staff.Role != nil {
		role = *body.Staff.Role
	}
	if role == "farm_admin" {
		exists, checkErr := h.hasActiveFarmAdmin(r.Context(), farmID, nil)
		if checkErr != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "직원 등록 중 오류가 발생했습니다."})
			return
		}
		if exists {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "농장관리자는 농장당 1명만 등록할 수 있습니다."})
			return
		}
	}
	empType := "full_time"
	if body.Staff.EmploymentType != nil {
		empType = *body.Staff.EmploymentType
	}
	userFarmID := uuid.New()
	_, err = h.db.Pool.Exec(r.Context(), `
		INSERT INTO user_farms (id, "userId", "farmId", role, department, position, "employmentType", "hireDate", "assignedBy", "assignedAt", "createdAt", "updatedAt")
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, NOW(), NOW(), NOW())
	`, userFarmID, userID, farmID, role, body.Staff.Department, body.Staff.Position, empType, body.Staff.HireDate, claims.UserID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "직원 등록 중 오류가 발생했습니다."})
		return
	}
	writeJSON(w, http.StatusCreated, map[string]interface{}{"user": map[string]interface{}{"id": userID.String(), "username": body.Account.Username, "fullName": body.Account.FullName}, "userFarm": map[string]interface{}{"id": userFarmID.String()}})
}

// FarmStaffUpdate PUT /api/farms/:farmId/staff/:userFarmId
func (h *Handler) FarmStaffUpdate(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r.Context())
	if claims == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "로그인이 필요합니다."})
		return
	}
	farmIDStr := chi.URLParam(r, "farmId")
	userFarmIDStr := chi.URLParam(r, "userFarmId")
	farmID, _ := uuid.Parse(farmIDStr)
	userFarmID, err := uuid.Parse(userFarmIDStr)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청입니다."})
		return
	}
	if !h.canManageFarmStaff(r.Context(), claims, farmID) {
		writeJSON(w, http.StatusForbidden, map[string]string{"error": "해당 농장의 직원을 관리할 권한이 없습니다."})
		return
	}
	var body struct {
		User  *struct {
			FullName *string `json:"fullName"`
			Phone   *string `json:"phone"`
			Email   *string `json:"email"`
		} `json:"user"`
		Staff *struct {
			Role           *string `json:"role"`
			Department     *string `json:"department"`
			Position       *string `json:"position"`
			EmploymentType *string `json:"employmentType"`
			HireDate       *string `json:"hireDate"`
			ResignDate     *string `json:"resignDate"`
		} `json:"staff"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청입니다."})
		return
	}
	var userId uuid.UUID
	var currentRole string
	if err := h.db.Pool.QueryRow(r.Context(), `SELECT "userId", role FROM user_farms WHERE id = $1 AND "farmId" = $2 AND "isActive" = true`, userFarmID, farmID).Scan(&userId, &currentRole); err != nil {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "해당 농장의 직원 정보를 찾을 수 없습니다."})
		return
	}
	if body.Staff != nil {
		if body.Staff.Role != nil && *body.Staff.Role == "farm_admin" && currentRole != "farm_admin" {
			exists, checkErr := h.hasActiveFarmAdmin(r.Context(), farmID, &userFarmID)
			if checkErr != nil {
				writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "수정 중 오류가 발생했습니다."})
				return
			}
			if exists {
				writeJSON(w, http.StatusBadRequest, map[string]string{"error": "농장관리자는 농장당 1명만 등록할 수 있습니다."})
				return
			}
		}
		_, _ = h.db.Pool.Exec(r.Context(), `UPDATE user_farms SET role = COALESCE($1, role), department = $2, position = $3, "employmentType" = COALESCE($4, "employmentType"), "hireDate" = $5, "resignDate" = $6, "updatedAt" = NOW() WHERE id = $7`,
			body.Staff.Role, body.Staff.Department, body.Staff.Position, body.Staff.EmploymentType, body.Staff.HireDate, body.Staff.ResignDate, userFarmID)
	}
	if body.User != nil {
		_, _ = h.db.Pool.Exec(r.Context(), `UPDATE users SET "fullName" = COALESCE($1, "fullName"), phone = $2, email = $3, "updatedAt" = NOW() WHERE id = $4`,
			body.User.FullName, body.User.Phone, body.User.Email, userId)
	}
	writeJSON(w, http.StatusOK, map[string]string{"message": "수정되었습니다."})
}

// FarmStaffDelete DELETE /api/farms/:farmId/staff/:userFarmId (soft: isActive=false)
func (h *Handler) FarmStaffDelete(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r.Context())
	if claims == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "로그인이 필요합니다."})
		return
	}
	farmIDStr := chi.URLParam(r, "farmId")
	userFarmIDStr := chi.URLParam(r, "userFarmId")
	farmID, _ := uuid.Parse(farmIDStr)
	userFarmID, err := uuid.Parse(userFarmIDStr)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청입니다."})
		return
	}
	if !h.canManageFarmStaff(r.Context(), claims, farmID) {
		writeJSON(w, http.StatusForbidden, map[string]string{"error": "해당 농장의 직원을 관리할 권한이 없습니다."})
		return
	}
	res, err := h.db.Pool.Exec(r.Context(), `UPDATE user_farms SET "isActive" = false, "resignDate" = COALESCE("resignDate", CURRENT_DATE::text), "updatedAt" = NOW() WHERE id = $1 AND "farmId" = $2`, userFarmID, farmID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "직원 삭제 처리 중 오류가 발생했습니다."})
		return
	}
	if res.RowsAffected() == 0 {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "해당 농장의 직원 정보를 찾을 수 없습니다."})
		return
	}
	writeJSON(w, http.StatusOK, map[string]interface{}{"success": true, "message": "퇴사 처리되었습니다."})
}

// FarmStaffLink POST /api/farms/:farmId/staff/link (existing user to farm)
func (h *Handler) FarmStaffLink(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r.Context())
	if claims == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "로그인이 필요합니다."})
		return
	}
	farmIDStr := chi.URLParam(r, "farmId")
	farmID, err := uuid.Parse(farmIDStr)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "farmId가 필요합니다."})
		return
	}
	if !h.canManageFarmStaff(r.Context(), claims, farmID) {
		writeJSON(w, http.StatusForbidden, map[string]string{"error": "해당 농장의 직원을 관리할 권한이 없습니다."})
		return
	}
	var body struct {
		UserID         string  `json:"userId"`
		Role           *string `json:"role"`
		Department     *string `json:"department"`
		Position       *string `json:"position"`
		EmploymentType *string `json:"employmentType"`
		HireDate       *string `json:"hireDate"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청입니다."})
		return
	}
	if body.UserID == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "userId는 필수입니다."})
		return
	}
	userID, err := uuid.Parse(body.UserID)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "userId가 올바르지 않습니다."})
		return
	}
	var exists int
	if err := h.db.Pool.QueryRow(r.Context(), `SELECT 1 FROM users WHERE id = $1`, userID).Scan(&exists); err != nil {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "해당 사용자를 찾을 수 없습니다."})
		return
	}
	if err := h.db.Pool.QueryRow(r.Context(), `SELECT 1 FROM user_farms WHERE "userId" = $1 AND "farmId" = $2`, userID, farmID).Scan(&exists); err == nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "이미 이 농장에 등록된 직원입니다."})
		return
	}
	role := "staff"
	if body.Role != nil {
		role = *body.Role
	}
	if role == "farm_admin" {
		exists, checkErr := h.hasActiveFarmAdmin(r.Context(), farmID, nil)
		if checkErr != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "기존 직원을 이 농장에 추가하는 중 오류가 발생했습니다."})
			return
		}
		if exists {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "농장관리자는 농장당 1명만 등록할 수 있습니다."})
			return
		}
	}
	empType := "full_time"
	if body.EmploymentType != nil {
		empType = *body.EmploymentType
	}
	userFarmID := uuid.New()
	_, err = h.db.Pool.Exec(r.Context(), `
		INSERT INTO user_farms (id, "userId", "farmId", role, department, position, "employmentType", "hireDate", "assignedBy", "assignedAt", "createdAt", "updatedAt")
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, NOW(), NOW(), NOW())
	`, userFarmID, userID, farmID, role, body.Department, body.Position, empType, body.HireDate, claims.UserID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "기존 직원을 이 농장에 추가하는 중 오류가 발생했습니다."})
		return
	}
	writeJSON(w, http.StatusCreated, map[string]interface{}{"userFarm": map[string]interface{}{"id": userFarmID.String()}})
}

// canManageFarmStaff: super_admin/system_admin or farm_admin of this farm
func (h *Handler) canManageFarmStaff(ctx context.Context, claims *middleware.Claims, farmID uuid.UUID) bool {
	if claims.SystemRole == "super_admin" || claims.SystemRole == "system_admin" {
		return true
	}
	var role string
	err := h.db.Pool.QueryRow(ctx, `SELECT role FROM user_farms WHERE "userId" = $1 AND "farmId" = $2 AND "isActive" = true`, claims.UserID, farmID).Scan(&role)
	return err == nil && role == "farm_admin"
}

func (h *Handler) hasActiveFarmAdmin(ctx context.Context, farmID uuid.UUID, excludeUserFarmID *uuid.UUID) (bool, error) {
	var exists bool
	var exclude interface{}
	if excludeUserFarmID != nil {
		exclude = *excludeUserFarmID
	}
	err := h.db.Pool.QueryRow(ctx, `
		SELECT EXISTS(
			SELECT 1
			FROM user_farms
			WHERE "farmId" = $1
			  AND "isActive" = true
			  AND role = 'farm_admin'
			  AND ($2::uuid IS NULL OR id <> $2)
		)
	`, farmID, exclude).Scan(&exists)
	if err != nil {
		return false, err
	}
	return exists, nil
}
